packages/ibm-products/src/components/Decorator/Decorator.js

```diff
@@ -0,0 +1,354 @@
+    if (typeof score !== 'number') {
+      return 'Unknown score';
+    }
+    return `"${magnitude}" magnitude. Score ${score} out of ${
+      scoreThresholds.slice(-1)[0]
+    }`;
+  },
+  scoreThresholds: [0, 4, 7, 10],
+  theme: 'light',
+};
+
+/**
+ * The Decorator groups a key/value pair to look and behave like a single UI element.
+ */
+export let Decorator = React.forwardRef(
+  (
+    {
+      // The component props, in alphabetical order (for consistency).
+      className,
+      disabled,
+      hideIcon,
+      href,
+      label,
+      labelTitle = defaults.labelTitle,
+      onClick,
+      onClickLabel,
+      onClickValue,
+      onContextMenu,
+      score,
+      scoreThresholds = defaults.scoreThresholds,
+      small,
+      theme = defaults.theme,
+      truncateValue,
+      value,
+      valueTitle,
+      // Collect any other property values passed in.
+      ...rest
+    },
+    ref
+  ) => {
+    const magnitude = useMemo(() => {
+      return getMagnitude(score, scoreThresholds);
+    }, [score, scoreThresholds]);
+
+    const _labelTitle = useMemo(() => {
+      return labelTitle && labelTitle(score, scoreThresholds, magnitude);
+    }, [magnitude, score, labelTitle, scoreThresholds]);
+
+    // If a "midline" truncation has been defined,
+    // then use the formatted midline value generated here,
+    // else refer to the original value.
+    const _value = useMemo(() => {
+      if (truncateValue?.maxLength) {
+        return truncate(
+          value,
+          truncateValue.maxLength,
+          truncateValue.front,
+          truncateValue.back
+        );
+      }
+
+      return value;
+    }, [truncateValue, value]);
+
+    // These class names apply to all types of Decorator.
+    const classNames = useMemo(() => {
+      return cx(blockClass, className, `${blockClass}--${theme}`, {
+        [`${blockClass}--sm`]: small,
+        [`${blockClass}--truncate-end`]: truncateValue === 'end',
+        [`${blockClass}--truncate-start`]: truncateValue === 'start',
+        [`${blockClass}--truncate-midline`]: truncateValue?.maxLength,
+      });
+    }, [className, small, theme, truncateValue]);
+
+    // These properties apply to all <DecoratorIcons>.
+    const decoratorIconsProps = {
+      className: `${blockClass}__icon`,
+      magnitude: magnitude.toLowerCase(), // e.g. "Medium" -> "medium"
+      small: small,
+    };
+
+    const handleOnClick = (event) => {
+      onClick(event, { score, label, value, magnitude });
+    };
+
+    const handleOnClickLabel = (event) => {
+      onClickLabel(event, { score, label, value, magnitude });
+    };
+
+    const handleOnClickValue = (event) => {
+      onClickValue(event, { score, label, value, magnitude });
+    };
+
+    const handleOnContextMenu = (event) => {
+      onContextMenu && onContextMenu(event, { score, label, value, magnitude });
+    };
+
+    // RETURN DUAL BUTTONS
+    if (onClickLabel && onClickValue) {
```

Comment:
i think a lesson we've learned over time is to try to make our components dumb when possible. while it might make sense to infer what variant the user needs based on props we tend to run into issues where because something isn't set directly users get confused about why they ended up with Y variant when they wanted X. to save us any headache i recommend removing these inferred conditions and just setting a new prop called `variant` or `type` or something and just having the user manually enter `single`, `dual`, etc

Reply:
I like it.

What better options, if any, are there for enforcing requirements? E.g. if kind="dual buttons", then onClickLabel and onClickValue must be defined as well. Can it be done through PropTypes? Should a console warning/error be logged? Throw an error (seems harsh)?

Reply:
we do have a utility for this kind of thing, though honestly i don't think it's super necessary to be this prescriptive. but if you so choose to there's a bunch of examples of it being used through the repo. just search for `.isRequired.if`

Reply:
`.isRequired.if` was a nice idea, but VSCode kept throwing an `"if" is not a function` error for my file alone. I verified against other examples, including looking for imports, but found no real difference. ¯\_(ツ)_/¯ 

I dropped it in favour of documentation.